File: src/compoundSwitch/CompoundSwitch.cpp
/*******************************************************************************
 * This file is part of openWNS (open Wireless Network Simulator)
 * _____________________________________________________________________________
 *
 * Copyright (C) 2004-2007
 * Chair of Communication Networks (ComNets)
 * Kopernikusstr. 16, D-52074 Aachen, Germany
 * phone: ++49-241-80-27910,
 * fax: ++49-241-80-22242
 * email: info@openwns.org
 * www: http://www.openwns.org
 * _____________________________________________________________________________
 *
 * openWNS is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 2 as published by the
 * Free Software Foundation;
 *
 * openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <DLL/compoundSwitch/CompoundSwitch.hpp>
#include <DLL/compoundSwitch/Filter.hpp>

#include <WNS/pyconfig/View.hpp>
#include <WNS/Assure.hpp>

using namespace dll::compoundSwitch;

STATIC_FACTORY_REGISTER_WITH_CREATOR(CompoundSwitch,
									 wns::ldk::FunctionalUnit,
									 "dll.compoundSwitch.CompoundSwitch",
									 wns::ldk::FUNConfigCreator);

CompoundSwitch::CompoundSwitch(wns::ldk::fun::FUN* fun, const wns::pyconfig::View& config) :
	wns::ldk::CommandTypeSpecifier<CompoundSwitchCommand>(fun),
	wns::ldk::HasReceptor<>(),
	wns::ldk::HasConnector<CompoundSwitchConnector>(),
	wns::ldk::HasDeliverer<CompoundSwitchDeliverer>(),
    wns::Cloneable<CompoundSwitch>(),
    friends_(),
	logger_(config.get("logger")),
    mustAccept_(config.get<bool>("mustAccept"))
{
	// configure all onData Filter
	int nOnDataFilter = config.len("onDataFilters");

	MESSAGE_SINGLE(NORMAL, logger_, "Configuring onDataFilters. Number of onDataFilters: " << nOnDataFilter);

	for ( int i = 0; i < nOnDataFilter; ++i )
	{
		wns::pyconfig::View onDataFilterConfig( config, "onDataFilters", i );
		std::string pluginName = onDataFilterConfig.get<std::string>("__plugin__");

		Filter* onDataFilter( FilterFactory::creator(pluginName)
							  ->create( this, onDataFilterConfig ) );

		getDeliverer()->addFilter(onDataFilter);
	}

	// configure all sendData Filter
	int nSendDataFilter = config.len("sendDataFilters");

    MESSAGE_SINGLE(NORMAL, logger_, "Configuring sendDataFilters. Number of sendDataFilters: " << nSendDataFilter);

	for ( int i = 0; i < nSendDataFilter; ++i )
	{
		wns::pyconfig::View sendDataFilterConfig( config, "sendDataFilters", i );
		std::string pluginName	= sendDataFilterConfig.get<std::string>("__plugin__");

		Filter* sendDataFilter( FilterFactory::creator(pluginName)
								->create( this, sendDataFilterConfig ) );

		getConnector()->addFilter(sendDataFilter);
	}

} // CompoundSwitch

CompoundSwitch::~CompoundSwitch()
{

} // ~CompoundSwitch

void
CompoundSwitch::onFUNCreated()
{
	getDeliverer()->onFUNCreated();
	getConnector()->onFUNCreated(mustAccept_);

	printFilterAssociation();
}

wns::ldk::FunctionalUnit*
CompoundSwitch::findFUNFriend(std::string friendName)
{
	return getFUN()->findFriend<FunctionalUnit*>(friendName);
}

////////// private functions ///////////////////////////////
void
CompoundSwitch::doSendData(const wns::ldk::CompoundPtr& compound)
{
	MESSAGE_SINGLE(NORMAL, logger_, "Outgoing compound catch by filter " << getConnector()->getFilter(compound)->getName());
	getConnector()->getAcceptor(compound)->sendData(compound);
} // doSendData

void
CompoundSwitch::doOnData(const wns::ldk::CompoundPtr& compound)
{
	MESSAGE_SINGLE(NORMAL, logger_, "Incomming compound catch by filter " << getDeliverer()->getFilter(compound)->getName());
	getDeliverer()->getAcceptor(compound)->onData(compound);
} // doOnData

bool
CompoundSwitch::doIsAccepting(const wns::ldk::CompoundPtr& compound) const
{
	return getConnector()->hasAcceptor(compound);

}

void
CompoundSwitch::doWakeup()
{
	getReceptor()->wakeup();
}

void
CompoundSwitch::printFilterAssociation()
{
    // onData
	MESSAGE_SINGLE(NORMAL, logger_, "onData association:   [ Filter -> Functional Unit ]");

	wns::ldk::Link::ExchangeContainer fus = getDeliverer()->get();
	CompoundSwitchDeliverer::Filters filters = getDeliverer()->getAllFilter();

	if ( filters.size() != fus.size() ) {
		std::stringstream ss;
		ss << "Configuration Error" << std::endl
		   << "Number of onData Filters mismatch number of upper FUs in Compound Switch" << std::endl
		   << "Number of onData Filters: " << filters.size() << " Number of upper FUs: " << fus.size() << std::endl;
		throw wns::Exception(ss.str());
	}

	CompoundSwitchDeliverer::Filters::iterator itFilter = filters.begin();
	for(wns::ldk::Link::ExchangeContainer::const_iterator itFU = fus.begin();
		itFU != fus.end();++itFU)
	{
		MESSAGE_SINGLE(NORMAL, logger_, (*itFilter)->getName() << " -> " << (*itFU)->getName());
		++itFilter;
	}

    //sendData
	MESSAGE_SINGLE(NORMAL, logger_,"sendData association: [ Filter, Functional Unit ]");

	fus = getConnector()->get();
	filters = getConnector()->getAllFilter();

	if ( filters.size() != fus.size() ) {
		std::stringstream ss;
		ss << "Configuration Error" << std::endl
		   << "Number of sendData Filters mismatch number of lower FUs in Compound Switch" << std::endl
		   << "Number of sendData Filters: " << filters.size() << " Number of lower FUs: " << fus.size() << std::endl;
		throw wns::Exception(ss.str());
	}

	itFilter = filters.begin();
	for(wns::ldk::Link::ExchangeContainer::const_iterator itFU = fus.begin();
		itFU != fus.end();++itFU)
	{
		MESSAGE_SINGLE(NORMAL, logger_, (*itFilter)->getName() << " -> " <<(*itFU)->getName());
		++itFilter;
	}

}

File: PyConfig/dll/CompoundSwitch.py
###############################################################################
# This file is part of openWNS (open Wireless Network Simulator)
# _____________________________________________________________________________
#
# Copyright (C) 2004-2007
# Chair of Communication Networks (ComNets)
# Kopernikusstr. 5, D-52074 Aachen, Germany
# phone: ++49-241-80-27910,
# fax: ++49-241-80-22242
# email: info@openwns.org
# www: http://www.openwns.org
# _____________________________________________________________________________
#
# openWNS is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License version 2 as published by the
# Free Software Foundation;
#
# openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################

import openwns.FUN
import openwns.logger
from openwns.pyconfig import attrsetter

class CompoundSwitch(openwns.FUN.FunctionalUnit):
    __plugin__="dll.compoundSwitch.CompoundSwitch"

    logger = None
    onDataFilters = None
    sendDataFilters = None
    mustAccept = None

    def __init__(self, functionalUnitName = 'compoundSwitch', commandName = 'compoundSwitch', logName = 'CompoundSwitch', moduleName = 'DLL', parentLogger = None, **kw):
        super(CompoundSwitch, self).__init__(functionalUnitName = functionalUnitName, commandName = commandName)
        self.logger =  openwns.logger.Logger(moduleName, logName, True, parentLogger)
        self.onDataFilters = []
        self.sendDataFilters = []
        self.mustAccept = True
        attrsetter(self, kw)

    def connectOnDataFU(self, fu, filter):
        self.onDataFilters.append(filter)
        fu.connect(self)

    def connectSendDataFU(self, fu, filter):
        self.sendDataFilters.append(filter)
        self.connect(fu)


class Filter(object):
    __plugin__="dll.compoundSwitch.Filter"
    name = None
    def __init__(self, name):
        self.name = name

######### filter implementation ###################
class FilterAll(Filter):
    __plugin__="dll.compoundSwitch.filter.FilterAll"

    def __init__(self,name,**kw):
        super(FilterAll, self).__init__(name)
        attrsetter(self, kw)


class FilterNone(Filter):
    __plugin__="dll.compoundSwitch.filter.FilterNone"

    def __init__(self,name,**kw):
        super(FilterNone, self).__init__(name)
        attrsetter(self, kw)


class FilterName(Filter):
    __plugin__="dll.compoundSwitch.filter.FilterName"

    def __init__(self,name,**kw):
        super(FilterFilterName, self).__init__(name)
        attrsetter(self, kw)


class FilterCommand(Filter):
    __plugin__="dll.compoundSwitch.filter.FilterCommand"
    commandProvider = None

    def __init__(self, name, commandProvider, **kw):
        super(FilterCommand, self).__init__(name)
        self.commandProvider = commandProvider
        attrsetter(self, kw)

class FilterMACAddress(Filter):
    __plugin__ = "dll.compoundSwitch.filter.FilterMACAddress"

    """ Allows for filtering by the MAC address, which was set by the upperConvergence"""

    upperConvergenceCommandName = None
    """ The name of the upperConvergence Command, 'upperConvergence' by default"""

    filterMACAddress = None
    """ The MAC addres the filter should match on """

    filterTarget = None
    """ If true, filters the target of the compound; if false, filters for the source"""

    def __init__(self, name, filterMACAddress, filterTarget, upperConvergenceCommandName, **kw):
        super(FilterMACAddress, self).__init__(name)
        self.upperConvergenceCommandName = upperConvergenceCommandName
        self.filterMACAddress = filterMACAddress
        self.filterTarget = filterTarget
        attrsetter(self, kw)
